connector/gitlab/gitlab.go

```diff
@@ -178,37 +186,39 @@ func (c *gitlabConnector) HandleCallback(s connector.Scopes, r *http.Request) (i
 }
 
 func (c *gitlabConnector) Refresh(ctx context.Context, s connector.Scopes, ident connector.Identity) (connector.Identity, error) {
-	if len(ident.ConnectorData) == 0 {
-		return ident, errors.New("no upstream access token found")
-	}
-
 	var data connectorData
 	if err := json.Unmarshal(ident.ConnectorData, &data); err != nil {
-		return ident, fmt.Errorf("gitlab: unmarshal access token: %v", err)
+		return ident, fmt.Errorf("gitlab: unmarshal connector data: %v", err)
 	}
+	oauth2Config := c.oauth2Config(s)
 
-	client := c.oauth2Config(s).Client(ctx, &oauth2.Token{AccessToken: data.AccessToken})
-	user, err := c.user(ctx, client)
-	if err != nil {
-		return ident, fmt.Errorf("gitlab: get user: %v", err)
+	if c.httpClient != nil {
+		ctx = context.WithValue(ctx, oauth2.HTTPClient, c.httpClient)
 	}
 
-	username := user.Name
-	if username == "" {
-		username = user.Email
+	t := &oauth2.Token{
```

Comment:
I propose to only use access token if there is no refresh token. Something like this:
```go
switch {
case data.RefreshToken != "":
    t := &oauth2.Token{
        RefreshToken: string(data.RefreshToken),
        t.Expiry: time.Now().Add(-time.Hour),
    }
    token, err = oauth2Config.TokenSource(ctx, t).Token()
	if err != nil {
		return ident, fmt.Errorf("gitlab: failed to get refresh token: %v", err)
	}
	return c.identity(ctx, s, token)
case data.AccessToken != "":
    t := &oauth2.Token{
       AccessToken: data.AccessToken,
    }
    return c.identity(ctx, s, t)
default:
	return ident, errors.New("no upstream refresh or access token found")
}
```
WDYT?

Reply:
Agreed, looking back on it, the initial check seems to weak to sufficiently tell when to retry the request with a the refresh token, I like this approach more.
I've changed it according to your suggestion as well as added a test when providing empty values for both `refreshToken` and `accessToken`, let me know if you feel the need for further changes.